src/scikit_build_core/file_api/_cattrs_converter.py

```diff
@@ -37,16 +44,33 @@ def make_converter(base_dir: Path) -> cattr.preconf.json.JsonConverter:
     converter.register_structure_hook(Reply, st_hook)
 
     def from_json_file(with_path: Dict[str, Any], t: Type[T]) -> T:
+        if "jsonFile" not in with_path and t is CodeModelDirectory:
+            return converter.structure_attrs_fromdict(with_path, t)
         if with_path["jsonFile"] is None:
             return converter.structure_attrs_fromdict({}, t)
         path = base_dir / Path(with_path["jsonFile"])
         raw = json.loads(path.read_text(encoding="utf-8"))
+        if t is CodeModelDirectory:
+            t = Directory  # type: ignore[assignment]
         return converter.structure_attrs_fromdict(raw, t)
 
+    def from_union(obj: Dict[str, Any], t: Type[T]) -> T:
+        for try_type in get_args(t):
+            try:
+                return converter.structure(obj, try_type)  # type: ignore[no-any-return]
+            except ClassValidationError:  # noqa: PERF203
+                continue
+        msg = f"Could not convert {obj} into {t}"
+        raise TypeError(msg)
+
     converter.register_structure_hook(CodeModel, from_json_file)
     converter.register_structure_hook(Target, from_json_file)
     converter.register_structure_hook(Cache, from_json_file)
     converter.register_structure_hook(CMakeFiles, from_json_file)
+    converter.register_structure_hook(CodeModelDirectory, from_json_file)
+    # Workaround for cattrs < 23.2.0 not handling Union with dataclass properly
+    if Version(version("cattrs")) < Version("23.2.0"):
```

Comment:
Should we just make this the minimum? It's a test-only dependency.

Reply:
From Fedora and epel we are fine, Debian and Ubuntu 24.04 seem to be on 22.2. They could skip and patch out the cattrs tests though, so maybe not a blocker

Reply:
It's fine, let's keep it as is. Going to update pyproject-metadata and make a patch release for #1006.